src/Elf.h

```diff
@@ -0,0 +1,395 @@
+    };
+
+    enum Flag : uint32_t {
+        SHF_WRITE = 0x1,
+        SHF_ALLOC = 0x2,
+        SHF_EXECINSTR = 0x4,
+        SHF_MASKPROC = 0xf0000000,
+    };
+
+    typedef std::vector<Relocation> RelocationList;
+    typedef RelocationList::iterator relocation_iterator;
+    typedef RelocationList::const_iterator const_relocation_iterator;
+
+    typedef std::vector<char>::iterator contents_iterator;
+    typedef std::vector<char>::const_iterator const_contents_iterator;
+
+private:
+    std::string name;
+    Type type = SHT_NULL;
+    uint32_t flags = 0;
+    std::vector<char> contents;
+    // Sections may have a size larger than the contents.
+    uint64_t size = 0;
+    uint64_t alignment = 1;
+    RelocationList relocs;
+
+public:
+    Section() {}
+    Section(const std::string &name, Type type) : name(name), type(type) {}
+
+    Section &set_name(const std::string &name) { this->name = name; return *this; }
+    const std::string &get_name() const { return name; }
+
+    Section &set_type(Type type) { this->type = type; return *this; }
+    Type get_type() const { return type; }
+
+    Section &set_flag(Flag flag) { this->flags |= flag; return *this; }
+    Section &remove_flag(Flag flag) { this->flags &= ~flag; return *this; }
+    Section &set_flags(uint32_t flags) { this->flags = flags; return *this; }
+    uint32_t get_flags() const { return flags; }
+    bool is_alloc() const { return (flags & SHF_ALLOC) != 0; }
+    bool is_writable() const { return (flags & SHF_WRITE) != 0; }
+
+    /** Get or set the size of the section. The size may be larger
+     * than the content. */
+    ///@{
+    Section &set_size(uint64_t size) { this->size = size; return *this; }
+    uint64_t get_size() const { return std::max(size, contents.size()); }
+    ///@}
+
+    Section &set_alignment(uint64_t alignment) { this->alignment = alignment; return *this; }
+    uint64_t get_alignment() const { return alignment; }
+
+    Section &set_contents(std::vector<char> contents) { this->contents = std::move(contents); return *this; }
+    template <typename It>
+    Section &set_contents(It begin, It end) { this->contents.assign(begin, end); return *this; }
+    template <typename It>
+    Section &append_contents(It begin, It end) {
+        this->contents.insert(this->contents.end(), begin, end);
+        return *this;
+    }
+    /** Set or append an object to the contents, assuming T is a
+     * trivially copyable datatype. */
+    template <typename T>
+    Section &set_contents(const std::vector<T> &contents) {
+        this->contents.assign((const char *)contents.data(), (const char *)(contents.data() + contents.size()));
+        return *this;
+    }
+    template <typename T>
+    Section &append_contents(const T& x) {
+        return append_contents((const char *)&x, (const char *)(&x + 1));
+    }
+    const std::vector<char> &get_contents() const { return contents; }
+    contents_iterator contents_begin() { return contents.begin(); }
```

Comment:
Is all this iterating functionality used? Might be simpler to just have contents()/relocations() return a reference to the underlying vector.